### ISSUE TITLE
migrate nextcloud from mariadb to postgres

### ISSUE THREAD

**Issue by tlake** [tlake/tlake-infra#36]
blocks #35

the `occ db:convert-type` command is currently disabled in my nextcloud version, so migrating from mariadb to postgres is on hold. it seems like other folks are running into the same issue: https://github.com/linuxserver/docker-nextcloud/issues/441

the issue above seemed to indicate that this would be resolved with the 29.0.4 release, but it looks like that version is already out and later it also references this issue https://github.com/nextcloud/server/issues/45257 which is slated for 29.0.6. we'll see.

there's also this comment https://github.com/nextcloud/server/issues/45257#issuecomment-2276730407 which suggests that i could manually patch, but i probably won't want to do so. i expect i'll just wait around until whatever maintenance patch comes out with the resolution.

**Comment by Investigamer** [comment]
Yeah some of us have been waiting for a long time on this (sqlite to postgres for me 😂). Looking like this won't be fixed till at least 29.0.7 I believe.

**Comment by tlake** [comment]
Nextcloud 30.0.1 is out and still no movement on the issue, so I've resorted to attempting the manual patch process. I'm running 29.0.8 and I followed the steps by @Delogon in [this comment](https://github.com/nextcloud/server/issues/45257#issuecomment-2363167396):

> For me the combined solutions of [@Queuecumber](https://github.com/Queuecumber), [@normen](https://github.com/normen) and [@lightonflux](https://github.com/lightonflux) worked, and I was able to convert my mariadb to postgresql.
> 
> So combined it would look like this:
> 
> In `lib/private/DB/ConnectionFactory.php`:
>
> ```
> -               $connectionParams = $this->createConnectionParams('', $additionalConnectionParams);
> +               $connectionParams = $this->createConnectionParams('', $additionalConnectionParams, $type);
> ```
>
> ```
> -       public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = []) {
> -               $type = $this->config->getValue('dbtype', 'sqlite');
> +       public function createConnectionParams(string $configPrefix = '', array $additionalConnectionParams = [], $type = '') {
> +               if ($type === '') {
> +                       $type = $this->config->getValue('dbtype', 'sqlite');
> +               }
> ```
>
> ```
> +               if ($type === '') {
> +                       $type = $this->config->getValue('dbtype', 'sqlite');
> +               }
> -               if ($this->config->getValue('mysql.utf8mb4', false)) {
> +               if ($this->config->getValue('mysql.utf8mb4', false) && $type==="mysql") {
> ```
>
> 
> In `core/Command/Db/ConvertType.php`
> 
> ```
> - $connectionParams = $this->connectionFactory->createConnectionParams();
> + $connectionParams = $this->connectionFactory->createConnectionParams('', [], $type);
> ```
>
> Maybe someone else could also check this?

I also had to disable the safety lock in `core/Command/Db/ConvertType.php`, of course:

```git
- throw new \InvalidArgumentException(
-         'This command is temporarily disabled (until the next maintenance release).'
- );
```

But, I'm pleased to report that the migration ran successfully and that I'm running with postgres now, so my thanks to everyone in that thread who contributed to this solution!
